File: googology/core/notation.py
from enum import IntEnum
from .capability import Op, Capabilities


class Family(IntEnum):
    """Top-level taxonomy of notations."""
    Ordinal = 0
    Number = 1
    Hierarchy = 2
    RealSequence = 3


class UnsupportedOperation(Exception):
    def __init__(self, name: str, op: Op):
        super().__init__(f"Notation '{name}' does not implement the requested operation")


class NotComparable(Exception):
    def __init__(self, name: str):
        super().__init__(f"Notation '{name}': comparison is undefined for large-number notations")


class Notation:
    """Generic notation base. Every notation derives from this class."""

    # --- taxonomy ---
    @property
    def name(self) -> str:
        raise NotImplementedError

    @property
    def family(self) -> Family:
        raise NotImplementedError

    @property
    def subfamily(self) -> str:
        return ""

    @property
    def style(self) -> str:
        return ""

    # --- attribution ---
    @property
    def creator(self) -> str:
        return ""

    @property
    def version(self) -> str:
        return ""

    # --- capability ---
    @property
    def capabilities(self) -> Capabilities:
        raise NotImplementedError

    def can(self, op: Op) -> bool:
        return self.capabilities.has(op)

    # --- string conversion ---
    def string_to_it(self, s: str) -> None:
        raise UnsupportedOperation(self.name, Op.FromString)

    def to_latex(self) -> str:
        raise NotImplementedError

    # --- operations ---
    def compare(self, other: 'Notation') -> int:
        raise NotComparable(self.name)

    def expand(self, n: int = 1) -> None:
        raise UnsupportedOperation(self.name, Op.Expand)

    def expand_to(self, length: int) -> None:
        raise UnsupportedOperation(self.name, Op.ExpandTo)

    @property
    def comparable(self) -> bool:
        return self.can(Op.Compare)

    def reduce(self) -> None:
        prev = self.to_latex()
        for _ in range(1000):
            self.expand(1)
            cur = self.to_latex()
            if cur == prev:
                return
            prev = cur

    def __str__(self) -> str:
        return self.to_latex()

File: src/core/notation.rs
use std::fmt;
use super::capability::{Op, Capabilities};

/// Top-level taxonomy of notations.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Family {
    Ordinal,
    Number,
    Hierarchy,
    RealSequence,
}

/// Error type for unsupported operations.
#[derive(Debug, Clone)]
pub struct UnsupportedOperation {
    pub name: String,
    pub op: Op,
}

impl std::fmt::Display for UnsupportedOperation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Notation '{}' does not implement the requested operation", self.name)
    }
}

impl std::error::Error for UnsupportedOperation {}

/// Error type for non-comparable notations (large-number notations).
#[derive(Debug, Clone)]
pub struct NotComparable {
    pub name: String,
}

impl std::fmt::Display for NotComparable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Notation '{}': comparison is undefined for large-number notations", self.name)
    }
}

impl std::error::Error for NotComparable {}

/// Generic notation base trait. Every notation implements this trait and
/// declares which operations it supports via capabilities().
///
/// The library NEVER computes a numeric value. expand()/expand_to()/reduce()
/// rewrite the notation and return the notation's own type.
pub trait Notation: fmt::Display {
    // --- taxonomy ---
    fn name(&self) -> &str;
    fn family(&self) -> Family;
    fn subfamily(&self) -> &str { "" }
    fn style(&self) -> &str { "" }

    // --- attribution ---
    fn creator(&self) -> &str { "" }
    fn version(&self) -> &str { "" }

    // --- capability ---
    fn capabilities(&self) -> Capabilities;
    fn can(&self, op: Op) -> bool {
        self.capabilities().has(op)
    }

    // --- string conversion ---
    fn string_to_it(&mut self, _s: &str) -> Result<(), String> {
        Err(UnsupportedOperation { name: self.name().to_string(), op: Op::FromString }.to_string())
    }
    fn to_string(&self) -> String;

    // --- operations ---
    fn compare(&self, _other: &dyn Notation) -> Result<i32, String> {
        Err(NotComparable { name: self.name().to_string() }.to_string())
    }
    fn expand(&mut self, _n: i64) -> Result<(), String> {
        Err(UnsupportedOperation { name: self.name().to_string(), op: Op::Expand }.to_string())
    }
    fn expand_to(&mut self, _len: i64) -> Result<(), String> {
        Err(UnsupportedOperation { name: self.name().to_string(), op: Op::ExpandTo }.to_string())
    }
    fn comparable(&self) -> bool {
        self.can(Op::Compare)
    }

    fn reduce(&mut self) -> Result<(), String> {
        let prev = self.to_string();
        for _ in 0..1000 {
            self.expand(1)?;
            let cur = self.to_string();
            if cur == prev {
                return Ok(());
            }
            // prev = cur; // would need to rebind — handled by loop
        }
        Ok(())
    }
}

impl fmt::Display for dyn Notation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.to_string())
    }
}